### DISCUSSION TITLE
processing argument to commands

### DISCUSSION THREAD

**Discussion by vlsireddy** [daniele77/cli#148] [Q&A] [answered]
Hi,

issue:
[1]   I used simplelocalsession.vcxproj example project  [Windows OS 64 bit MSVC 2022].
[2]  Auto completion is not working? is it supposed to work ?

question:
[1] how to handle MENU Commands with variable arguments like this
hello -time 1000 -flow 2000 -stream 3000 -Parallel yes
hello --time 1000 --flow 2000 --stream 3000 -Parallel yes
hello -t 1000 -f 2000 -s 3000 -P y
I tried above commands it lands into following error
wrong command: <above commands>
Is it possible with your cli framework ?

Thanks

**Comment by daniele77** [comment, marked as answer]
Hi @vlsireddy ,
first of all, thank you for using cli library.

The `simplelocalsession` is an example of using a `SimpleLocalSession` that does not provide autocompletion and other valuable features.
Instead, you should use `CliLocalSession` (see the examples `complete` or `pluginmanager`).

Cli library is not for parsing application parameters, instead, it provides an interactive shell where you can enter commands in a positional way.
Every command has this syntax:
  
    <command_name> <parameter1> <parameter2> <parameter3> ...

So, instead of:

    hello -time 1000 -flow 2000 -stream 3000 -Parallel yes

you can have instead:

    hello 1000 2000 3000 true

**Comment by vlsireddy** [comment]
Hi,
    Thanks for reply. 

understood both points.

btw gr8 work.

Thanks

**Comment by suzizecat** [comment]
After making an off-topic reply on #69, this seems to be a more appropriate place to make my suggestion. 
TL;DR : You can bind a generic argument parser such as https://github.com/p-ranav/argparse to do the job.

While CLI does not provide a way to parse and handle arguments, it is possible to have an arbitrary list of arguments as a list of string. 
This matches the use case of a typical argument parser and therefore I tried to use one (argparse) with some success.

There is two tweaks to do with regard to a normal use of argparse:
 - Argparse expects to have the name of the executable as the first value in the argument list.
 - When using the default provided `--help` argument, argparse calls `std::exit(0)` which exit the whole program and produce weird behavior with CLI.

Fortunately, it is possible to work around those two issues relatively easily. 
To address the first point, the user only have to prepend an element (anything actually) to the vector of arguments.
For the second one, it is possible to override the `--help` argument, this is described in argparse README.

An example of this manipulation, is this code:
```cpp
static void cmd_read_lib(std::ostream& out, const std::vector<std::string>& _args){
	std::vector<std::string> extended_args = {"read"};
	std::copy(std::begin(_args),std::end(_args),std::back_inserter(extended_args));

	argparse::ArgumentParser args("Liberty reader","1.0.0",argparse::default_arguments::none);
	// Note default_arguments::none

	args.add_argument("path")
		.help("Liberty file path");

	args.add_argument("--threads","-j")
			.help("Number of threads to use. 0 for maximum")
			.scan<'i', int>()
			.default_value(1);
	args.add_argument("-h", "--help") 
			.action([=](const std::string& s) {
				std::cout << args; // Still display the help, but doesn't quit the program.
			})
			.default_value(false)
			.help("shows help message")
			.implicit_value(true)
			.nargs(0);
	args.parse_args(extended_args);

	if (args.get<bool>("help")) // Properly exit the function if help is requested.
		return;
	// Rest of the function
}

// ... In your menu building : 
function<void(ostream&,const vector<string>&)> f_read = cmd_read_lib;
root->Insert("read",f_read,"Read the liberty file");
```

When using the help, the argument parser will complain that there is a missing positional argument, but this is a minor issue and doesn't cause actual issue. 

Hoping that it can help !
